packages/shared/src/beacon.ts

```diff
@@ -0,0 +1,105 @@
+/* -----------------------------------
+ *
+ * sendBeacon with XMLHttpRequest
+ *
+ * -------------------------------- */
+
+const syncEvents = new Set(['unload', 'beforeunload', 'pagehide']);
+
+function sendBeaconXHR(url: string | URL, data?: XMLHttpRequestBodyInit | null): boolean {
+   const eventType = this?.event?.type;
+   const sync = syncEvents.has((eventType || '').toLowerCase());
+
+   const xhr = new XMLHttpRequest();
+   xhr.open('POST', url, !sync);
+   xhr.withCredentials = true;
+   xhr.setRequestHeader('Accept', '*/*');
+
+
+   if (typeof data === 'string') {
+      xhr.setRequestHeader('Content-Type', 'text/plain;charset=UTF-8');
+      xhr.responseType = 'text';
+   } else if ((data instanceof Blob) && data.type) {
+      xhr.setRequestHeader('Content-Type', data.type);
+   }
+
+   try {
+      xhr.send(data);
+   } catch (error) {
+      return false;
+   }
+
+   return true;
+}
+
+/* -----------------------------------
+ *
+ * sendBeacon with Fetch
+ *
+ * -------------------------------- */
+
+function sendBeaconFetch(url: string | URL, data?: XMLHttpRequestBodyInit | null): boolean {
```

Comment:
**suggestion**: Although we're not making use of the `Promise` returned via `fetch()` here (yet), it's still probably worth returning one here for future sake. Can we rework this so we can return & resolve the Promise used within this functions scope?

Reply:
> Good work! This looks great 👍
> 
> Just a few initial comments, and I believe you'll need to either run `eslint . --fix` or setup format on save for your IDE using ESLint & Prettier to get the formatting just right.
> 
> I'm also going to have a think about any other opportunities we might have to reduce the output file size a bit.

Thanks for the suggestions, I (hopefully) brought test coverage high enough now that it'll pass CI. Returning and passing through `Promise` will be a larger refactor that I'll get to when time permits.